File: infra/lib/creativeiq-stack.ts
// CreativeIqStack — DEFERRED stub for cloud deploy (prototype runs locally; see infra/README.md).
//
// When activated, this stack provisions: an S3 bucket for creatives + hero set, four Python 3.12
// Lambdas wrapping backend/handlers/*.py, an HTTP API Gateway, and least-privilege IAM
// (bedrock:InvokeModel[WithResponseStream] + S3 on the single bucket only). Streaming /generate
// uses Lambda response streaming. The skeleton below shows the intended shape.
import * as cdk from "aws-cdk-lib";
import { Construct } from "constructs";
// import * as s3 from "aws-cdk-lib/aws-s3";
// import * as lambda from "aws-cdk-lib/aws-lambda";
// import * as apigw from "aws-cdk-lib/aws-apigatewayv2";
// import * as iam from "aws-cdk-lib/aws-iam";

export class CreativeIqStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    // const bucket = new s3.Bucket(this, "CreativesBucket", {
    //   blockPublicAccess: s3.BlockPublicAccess.BLOCK_ALL,
    //   removalPolicy: cdk.RemovalPolicy.DESTROY, // demo only
    // });

    // const bedrockPolicy = new iam.PolicyStatement({
    //   actions: ["bedrock:InvokeModel", "bedrock:InvokeModelWithResponseStream"],
    //   resources: ["*"], // scope to specific model ARNs in production
    // });

    // const makeFn = (name: string, handler: string) =>
    //   new lambda.Function(this, name, {
    //     runtime: lambda.Runtime.PYTHON_3_12,
    //     code: lambda.Code.fromAsset("../backend"),
    //     handler, // e.g. "handlers.generate.lambda_handler"
    //     timeout: cdk.Duration.seconds(30),
    //     environment: { USE_BEDROCK: "1", CREATIVEIQ_REGION: this.region },
    //   });

    // const generateFn = makeFn("GenerateFn", "handlers.generate.lambda_handler");
    // ... predict / segments / catalog ...
    // bucket.grantReadWrite(generateFn);
    // generateFn.addToRolePolicy(bedrockPolicy);

    // const api = new apigw.HttpApi(this, "CreativeIqApi");
    // api.addRoutes({ path: "/generate", methods: [apigw.HttpMethod.POST], integration: ... });

    // new cdk.CfnOutput(this, "ApiBaseUrl", { value: api.apiEndpoint });

    new cdk.CfnOutput(this, "Status", {
      value: "DEFERRED: implement resources above to deploy. Prototype runs locally (see repo README).",
    });
  }
}

File: infra/cdk/creative_iq_stack.py
"""
CreativeIQ CDK stack (mirrors the CarbonClarity / DriveScore pattern).

  * Lambda (Python 3.12, x86_64) running the FastAPI backend via Mangum behind a
    Function URL (no Docker — Linux wheels). Bundles the synthetic data + the
    pre-generated hero images, served by the app's /data route.
  * Image generation uses the keyless text-to-image endpoint in DIRECT mode
    (CREATIVEIQ_IMAGE_DIRECT=1) so a cache-miss returns the image URL for the
    browser to load — no writes to the read-only Lambda filesystem.
  * IAM: bedrock:InvokeModel* (used only when USE_BEDROCK=1).
  * Frontend: private S3 + CloudFront (OAC, HTTPS) serving frontend/dist, built
    with VITE_API_BASE = VITE_ASSET_BASE = the Function URL (so /api calls and
    /data images both resolve to the Lambda).

Personal account <APP_ACCOUNT> / ap-southeast-1 only (see app.py guardrail +
scripts/deploy.sh account check).
"""
import os

from aws_cdk import (
    Stack, Duration, CfnOutput, RemovalPolicy,
    aws_lambda as _lambda,
    aws_iam as iam,
    aws_s3 as s3,
    aws_s3_deployment as s3deploy,
    aws_cloudfront as cf,
    aws_cloudfront_origins as origins,
)
from constructs import Construct

REGION = "ap-southeast-1"


class CreativeIqStack(Stack):
    def __init__(self, scope: Construct, cid: str, **kwargs) -> None:
        super().__init__(scope, cid, **kwargs)

        backend_asset = os.environ.get("BACKEND_ASSET", "../../build/backend")
        frontend_asset = os.environ.get("FRONTEND_ASSET", "../../frontend/dist")

        fn = _lambda.Function(
            self, "Api",
            runtime=_lambda.Runtime.PYTHON_3_12,
            architecture=_lambda.Architecture.X86_64,
            handler="lambda_function.handler",
            code=_lambda.Code.from_asset(backend_asset),
            timeout=Duration.seconds(60),
            memory_size=1024,
            environment={
                "CREATIVEIQ_DATA_DIR": "/var/task/data",
                "CREATIVEIQ_IMAGE_DIRECT": "1",
                "USE_BEDROCK": os.environ.get("USE_BEDROCK", "0"),
                "CREATIVEIQ_REGION": REGION,
            },
        )

        fn.add_to_role_policy(iam.PolicyStatement(
            actions=["bedrock:InvokeModel", "bedrock:InvokeModelWithResponseStream"],
            resources=[
                "arn:aws:bedrock:*::foundation-model/anthropic.*",
                "arn:aws:bedrock:*::foundation-model/amazon.*",
                f"arn:aws:bedrock:*:{self.account}:inference-profile/*",
            ],
        ))

        furl = fn.add_function_url(auth_type=_lambda.FunctionUrlAuthType.NONE)

        site = s3.Bucket(
            self, "Site",
            block_public_access=s3.BlockPublicAccess.BLOCK_ALL,
            removal_policy=RemovalPolicy.DESTROY,
            auto_delete_objects=True,
            enforce_ssl=True,
        )

        dist = cf.Distribution(
            self, "Cdn",
            default_behavior=cf.BehaviorOptions(
                origin=origins.S3BucketOrigin.with_origin_access_control(site),
                viewer_protocol_policy=cf.ViewerProtocolPolicy.REDIRECT_TO_HTTPS,
            ),
            default_root_object="index.html",
            error_responses=[
                cf.ErrorResponse(http_status=403, response_http_status=200, response_page_path="/index.html"),
                cf.ErrorResponse(http_status=404, response_http_status=200, response_page_path="/index.html"),
            ],
            comment="CreativeIQ static site",
        )

        s3deploy.BucketDeployment(
            self, "Deploy",
            sources=[s3deploy.Source.asset(frontend_asset)],
            destination_bucket=site,
            distribution=dist,
            distribution_paths=["/*"],
        )

        CfnOutput(self, "FunctionUrl", value=furl.url)
        CfnOutput(self, "SiteUrl", value=f"https://{dist.distribution_domain_name}")
        CfnOutput(self, "BucketName", value=site.bucket_name)
